dash/testing/browser.py

```
@@ -55,10 +58,18 @@ def __init__(
         )
         self.percy_runner.initialize_build()
 
+        logger.debug("initialize browser with arguments")
+        logger.debug("  headless => %s", self._headless)
+        logger.debug("  download_path => %s", self._download_path)
+
     def __enter__(self):
         return self
 
     def __exit__(self, exc_type, exc_val, traceback):
+        try:
+            shutil.rmtree(self._download_path)
+        except OSError:
+            logger.exception("deleting tmp download folder has an issue")
```

should we also raise an error if `self._download_path` already exists when we start? Would hate to have our framework delete something it wasn't supposed to!

---

OIC, in `dash_duo` and `dash_br`:
```py
download_path=tmpdir.mkdir('download').strpath
```

Feels weird to me - perhaps we should just have `Browser` make this dir automatically the first time the test asks for `dash_duo.download_path`, and then at the end only delete it if it was created.

---

hmm, probably I should just leave `pytest` to manage the lifecycle of the folder 
quote  ```tmpdir fixture provides a temporary directory unique to the test invocation, created in the base temporary directory.```

https://docs.pytest.org/en/latest/tmpdir.html#base-temporary-directory

---

> Feels weird to me - perhaps we should just have `Browser` make this dir automatically the first time the test asks for `dash_duo.download_path`, and then at the end only delete it if it was created.

it's agnostic at browser creation whether the fixture will be used for download.

---

ah that's right, the user wouldn't have any reason to access `download_path` until *after* the download has happened, which is of course too late to create it!

OK, so the way you have it now (after https://github.com/plotly/dash/pull/818/commits/998346656eab020019dc50da9d6e9d7dfe767bf8) the `tmpdir` fixture *always* makes this dir, and properly cleans it up afterward? As long as that's known to be reasonably robust across OSes, and the time overhead is minimal, seems good to me!

---

from the above link, `entries older than 3 temporary directories will be removed` so I think it's safer to rely on `pytest` than rebuilding the wheel.